DS-3910 Delete method on item endpoint

ppmdo opened the pull request:
Implemented DELETE method on the item endpoint.
Wrote IntegrationTests.
Wrote WorkspaceItemBuilder class.

Commit:
Added method

Commit:
Working method to DELETE on /api/core/items
Wrote integration tests

Commit:
Working Implementation
Wrote additional integration tests
Created WorkspaceItemBuilder

Commit:
WorskspaceItemBuilder amend

abollini commented on dspace-spring-rest/src/test/java/org/dspace/app/rest/builder/WorkspaceItemBuilder.java:
```diff
@@ -0,0 +1,66 @@
+import org.dspace.content.WorkspaceItem;
+import org.dspace.content.service.DSpaceObjectService;
+import org.dspace.core.Context;
+import org.dspace.eperson.Group;
```
this is not used and lead to the checkstyle violation reported by travis, see https://travis-ci.org/DSpace/DSpace/builds/399324480#L5632

abollini commented on dspace-spring-rest/src/test/java/org/dspace/app/rest/builder/WorkspaceItemBuilder.java:
```diff
@@ -0,0 +1,66 @@
+ * Builder to construct WorkspaceItem objects
+ *
+ **/
+public class WorkspaceItemBuilder extends AbstractDSpaceObjectBuilder<Item> {
```
This should extends AbstractDSpaceObjectBuilder<WorkspaceItem>

abollini replied:
You are right. So you should start from a lower level in the hierarchy org.dspace.app.rest.builder.AbstractCRUDBuilder<T> or org.dspace.app.rest.builder.AbstractBuilder<T, S>

abollini commented on dspace-spring-rest/src/test/java/org/dspace/app/rest/builder/WorkspaceItemBuilder.java:
```diff
@@ -0,0 +1,66 @@
+    @Override
+    public Item build() {
+        try {
+            return item;
```
a workspaceitem should be returned

ppmdo replied:
Done :)

abollini commented on dspace-spring-rest/src/test/java/org/dspace/app/rest/builder/WorkspaceItemBuilder.java:
```diff
@@ -0,0 +1,66 @@
+    }
+
+    protected void cleanup() throws Exception {
+        delete(item);
```
the workspaceitem needs to be deleted

ppmdo replied:
Done

abollini commented on dspace-spring-rest/src/test/java/org/dspace/app/rest/builder/ItemBuilder.java:
```diff
@@ -37,6 +37,20 @@ public static ItemBuilder createItem(final Context context, final Collection col
         return builder.create(context, col);
     }
 
+    public static Item createTemplateItem(final Context context, final Collection collection) {
```
this should go in the CollectionBuilder

ppmdo replied:
Done

abollini commented on dspace-spring-rest/src/test/java/org/dspace/app/rest/ItemRestRepositoryIT.java:
```diff
@@ -308,7 +307,112 @@ public void findOneTestWrongUUID() throws Exception {
+        String token = getAuthToken(admin.getEmail(), password);
+
+        //Check templateItem creation
+        getClient().perform(get("/api/core/items/" + templateItem.getID()))
```
it is more useful to test that the item is still here after a failed attempt to delete it

abollini commented on dspace-spring-rest/src/main/webapp/META-INF/context.xml:
```diff
@@ -0,0 +1,2 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
please remove this file. The context path is defined at the deploy time

abollini commented on dspace-spring-rest/src/main/resources/application.properties:
```diff
@@ -22,7 +22,7 @@
 #
 # DSpace installation directory
-dspace.dir=${dspace.dir}
+dspace.dir=/opt/dspace
```
please remove this local configuration

abollini commented on dspace-spring-rest/src/main/resources/application.properties:
```diff
@@ -77,7 +77,7 @@ server.port=8080
 # This creates a Tomcat context-param named "dspace.dir"
 # and sets it to the value of the "dspace.dir" property (listed above)
-server.context-parameters.dspace.dir=${dspace.dir}
+server.context-parameters.dspace.dir=/opt/dspace
```
please remote this local configuration

ppmdo replied:
Sorry for the local configurations. I'll remove them from the PR.

abollini commented on dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/ItemRestRepository.java:
```diff
@@ -88,4 +90,28 @@ public ItemResource wrapResource(ItemRest item, String... rels) {
+        }
+        try {
+            is.delete(context, item);
+            is.removeAllBundles(context, item);
```
this is not needed as done internally in the ItemService.delete
https://github.com/DSpace/DSpace/blob/master/dspace-api/src/main/java/org/dspace/content/ItemServiceImpl.java#L649

ppmdo replied:
Done

abollini reviewed (changes requested):
Hi Pablo,
thanks for your PR. Please check my comments some changes are required you are close to complete the task

abollini commented:
@ppmdo  I still don't see any changes in the PR, maybe you are not pushed your local changes to the branch from where this PR was created?
Please note also the comment about the AbstractDSpaceObjectBuilder

ppmdo commented:
Hi @abollini. Thanks for your comments. In fact I was awaiting your advise on de WorkspaceItemBuilder issue. I'll get this done asap. 👍

Commit:
Moved createTemplateItem method to CollectionBuilder class.
Correction: WorkspaceItemBuilder extends AbstractBuiilder

Commit:
Removed local configurations

Commit:
Merge branch 'DS-3910' of https://github.com/ppmdo/DSpace into DS-3910

Commit:
Revert "Removed local configurations"

This reverts commit 71347a3faf662930270b736e089a71d044edbad9.

Commit:
Amend

ppmdo commented:
@abollini I've just updated the PR :+1:

Commit:
Style corrections

tomdesair commented on dspace-spring-rest/src/test/java/org/dspace/app/rest/builder/CollectionBuilder.java:
```diff
@@ -78,6 +79,20 @@ public Collection build() {
         return collection;
     }
 
+    public static Item createTemplateItem(final Context context, final Collection collection) {
```
Can you move this method to the `ItemBuilder` class? That should allow us to create collection template items with metadata:

```
Item templateItem = ItemBuilder.createTemplateItemForCollection(context, col1)
                                                    .withSubject("Mathematics")
                                                    .build();
```

In the `createTemplateItemForCollection` you need to set a flag on the `ItemBuilder` object (e.g. `archiveItem = false;`) to remember that we should not call `installItem` on executing `build()`.

ppmdo replied:
Hi @tomdesair. Thanks for the review. @abollini I had this method in ItemBuilder class originally but @abollini  requested me to move to it CollectionBuilder. ¿How shall we proceed?

abollini replied:
the templateItem is really a piece of the collection (like a logo) so conceptually I think is more appropriate to have it created in the collectio builder.
Moreover, due to how the builders are call during the cleanup phase we could end with the collectionbuilder removing the templateitem and the itembuilder failing as the item doesn't longer exist.
Having the template item managed/known by a single builder is imho easier.

@tomdesair To add metadata to the template item can we simply get it after the collection creation (.build())?

tdonohue replied:
I agree with @abollini here.  This method conceptually should remain in the `CollectionBuilder`, as Item Templates are *not* normal Items. They belong to Collections, and can only be created from Collections.

If we find a reason to move this method in the future (to support future features/code), then we can move it at that time. For now though, I think we should just leave it where it is.

tdonohue replied:
@ppmdo : could you add Javadocs to this `public` method?  I notice it's missing Javadocs. Per our code style, we recommend Javadocs on all public methods/classes: https://wiki.duraspace.org/display/DSPACE/Code+Style+Guide

tomdesair commented on dspace-spring-rest/src/test/java/org/dspace/app/rest/builder/ItemBuilder.java:
```diff
@@ -37,7 +37,7 @@ public static ItemBuilder createItem(final Context context, final Collection col
     }
 
-    private ItemBuilder create(final Context context, final Collection col) {
+     private ItemBuilder create(final Context context, final Collection col) {
```
Can you remove this extra space again?

tomdesair commented on dspace-spring-rest/src/test/java/org/dspace/app/rest/builder/WorkspaceItemBuilder.java:
```diff
@@ -0,0 +1,86 @@
+        return this;
+    }
+
+    protected <B> B handleException(final Exception e) {
```
Could you move the `org.dspace.app.rest.builder.AbstractDSpaceObjectBuilder#handleException` to the `org.dspace.app.rest.builder.AbstractBuilder` class? That why you don't have to implement this method yourself and you can reuse the existing code.

tomdesair commented on dspace-spring-rest/src/test/java/org/dspace/app/rest/builder/WorkspaceItemBuilder.java:
```diff
@@ -0,0 +1,86 @@
+    public WorkspaceItem build() {
+        try {
+            return workspaceItem;
+        } catch (Exception e) {
```
The return statement can never throw an exception here. So please remove the try-catch block.

tomdesair commented on dspace-spring-rest/src/test/java/org/dspace/app/rest/builder/WorkspaceItemBuilder.java:
```diff
@@ -0,0 +1,86 @@
+    }
+
+    protected <B> B handleException(final Exception e) {
+        log.error(e.getMessage(), e);
```
You can also move the `log` variable to the `AbstractBuilder` class so that you don't have to create your own `log` variable.

tomdesair reviewed (commented):
Code looks good, nicely done!

I only have some minor (non-blocking) code-change remarks.

Commit:
Minor changes based on review.

ppmdo committed:
Merge branch 'master' into DS-3910

tdonohue reviewed (approved):
👍 The code here looks good to me.  However, I noticed there's a public method missing Javadocs. I'd appreciate it if we could add Javadocs before merging.

ppmdo commented:
Will do.

Commit:
Added JavaDoc for createTemplateItem

Commit:
Merge branch 'DS-3910' of https://github.com/ppmdo/DSpace into DS-3910

ppmdo commented:
@tdonohue Done

abollini commented on dspace-spring-rest/src/test/java/org/dspace/app/rest/builder/CollectionBuilder.java:
```diff
@@ -78,6 +79,26 @@ public Collection build() {
+    * @param context DSpace Context
+    * @param collection DSpace Collection
+    */
+    public static Item createTemplateItem(final Context context, final Collection collection) {
```
I'm not sure that we want to allow the use of a builder to modify the collection after the initial build. To be consistent with the existent code I would expect a withTemplateItem() method instead.

ppmdo replied:
I agree we should be consistent all around on how the code is implemented.

abollini commented on dspace-spring-rest/src/test/java/org/dspace/app/rest/builder/CollectionBuilder.java:
```diff
@@ -78,6 +79,26 @@ public Collection build() {
+
+    private Item create_template_item(final Collection collection) {
+        try {
+            Item item = itemService.createTemplateItem(context, collection);
```
If we decide to keep this method as is (see comment above) the builder should keep a reference to the created templateItem and remove it in the cleanup. It is only needed if this method is invoked on a "test collection" created outside the test method (that is the case of the basic test structure) otherwise the "wide test collection" will remain altered for the subsequent tests in the class.

ppmdo replied:
So, if we proceed with the withTemplateItem() the template item should be an attribute of the created collection and cease to exist upon the cleanup of the aforementioned collection. Right?

Commit:
Merge branch 'master' of https://github.com/DSpace/DSpace into DS-3910

Commit:
Changed createTemplateItem to collection.withTemplateItem() to keep code coherency.
Applied changes to ItemRestRepositoryIT to work using withTemplateItem

ppmdo commented:
@abollini I've just changed the createTemplateMethod to withTemplateItem as you pointed. This keeps 
consistency with the rest of the builders. 

The only issue we have to resolve is the exception handler. 

WorkspaceItemBuilder inherits directly from AbstractBuilder which doesn't have a handleException method, as AbstractDSpaceObjectBuilder does. That's why I implemented that method on that class. If you consider that method must be moved to another place o not used, just let me know and I'll change it.

terrywbrady reviewed (approved):
These changes look good to me.

If additional changes are needed for the handleException() method perhaps that could be tackled with a new ticket.

tdonohue commented:
I'm going to merge these changes as-is, since this PR is at +2 votes, and @ppmdo should be able to move along to other code/PRs :)

I agree with @terrywbrady that we can resolve the `exceptionHandler()` question in a followup PR as needed.  There are questions in general about best practices in terms of Exception handling in the REST API layer (see also comments under https://github.com/DSpace/DSpace/pull/2126#pullrequestreview-143324207 relating to using `RuntimeException` for exceptions).

So, I recommend we merge this PR and have a discussion about REST API exception handling when @abollini returns from travel.  We can then follow-up with fixes to exception handling based on whatever best practices we decide upon.
